nwb_conversion_tools/datainterfaces/behavior/movie/movie_utils.py

```diff
@@ -13,6 +15,134 @@
+        self.frame = self.get_movie_frame(0)
+        assert self.frame is not None, "unable to read the movie file provided"
+
+    def get_movie_timestamps(self):
+        """
+        Return numpy array of the timestamps for a movie file.
+
+        """
+        if not self.vc.isOpened():
+            raise ValueError("movie file is not open")
+        ts = [self.vc.get(cv2.CAP_PROP_POS_MSEC)]
+        for i in tqdm(range(1, self.get_movie_frame_count()), desc="retrieving video timestamps"):
+            self.current_frame = i
+            ts.append(self.vc.get(cv2.CAP_PROP_POS_MSEC))
+        self.current_frame = 0
+        return np.array(ts)
+
+    def get_movie_fps(self):
+        """
+        Return the internal frames per second (fps) for a movie file.
+
+        """
+        if int(cv2.__version__.split(".")[0]) < 3:
+            return self.vc.get(cv2.cv.CV_CAP_PROP_FPS)
+        return self.vc.get(cv2.CAP_PROP_FPS)
+
+    def get_frame_shape(self) -> Tuple:
+        """
+        Return the shape of frames from a movie file.
+        """
+        return self.frame.shape
+
+    def get_movie_frame_count(self):
+        """
+        Return the total number of frames for a movie file.
+
+        """
+        if self.stub:
+            # if stub the assume a max frame count of 10
+            return 10
+        if int(cv2.__version__.split(".")[0]) < 3:
+            count = self.vc.get(cv2.cv.CV_CAP_PROP_FRAME_COUNT)
+        else:
+            count = self.vc.get(cv2.CAP_PROP_FRAME_COUNT)
+        return int(count)
+
+    @property
+    def current_frame(self):
+        return self._current_frame
+
+    @current_frame.setter
+    def current_frame(self, frame_no):
+        if int(cv2.__version__.split(".")[0]) < 3:
+            set_arg = cv2.cv.CV_CAP_PROP_POS_FRAMES
+        else:
+            set_arg = cv2.CAP_PROP_POS_FRAMES
+        set_value = self.vc.set(set_arg, frame_no)
+        if set_value:
+            self._current_frame = frame_no
+        else:
+            raise ValueError(f"could not set frame no {frame_no}")
+
+    def get_movie_frame(self, frame_no: int):
+        """
+        Return the specific frame from a movie.
+        """
+        if not self.vc.isOpened():
+            raise ValueError("movie file is not open")
+        assert frame_no < self.get_movie_frame_count(), "frame number is greater than length of movie"
+        self.current_frame = frame_no
+        success, frame = self.vc.read()
+        self.current_frame = 0
+        if success:
+            return frame
+        elif frame_no > 0:
+            return np.nan * np.ones(self.get_frame_shape())
+
+    def get_movie_frame_dtype(self):
+        """
+        Return the dtype for frame in a movie file.
+        """
+        return self.frame.dtype
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if not self.vc.isOpened():
+            self.vc = cv2.VideoCapture(*self._args, **self._kwargs)
```

Comment:
What is the intent of forcing the file opening here? Might be best to leave definitions of the `self.vc` for the `__enter__` and `__init__` stages, and assume that the `next` is only intentionally called after doing one of those two entry points.

Reply:
I wanted this be an iterable even if the object is closed: reinstantiating the `VideoWriter` attribute. If its open (we are iterating after instantiating the class) then we can skip this re-instantiation. 
Moreover, `__next__` is releasing it each time so we would need to instantiate it. 
But now that I think of it, it might be expensive to reinstantiate after every call to `__next__` since it releases it. The `StopIteration` would hit before releasing it would be necessary. 
It might be better to not release it after every call and do so only before `StopIteration`. Then we can also just enforce the `VideoWriter` to be open before iterating over it. https://github.com/catalystneuro/nwb-conversion-tools/pull/392/commits/48db3f541b95b4fc41679a1e31f66eefa97290ed

Reply:
I would agree. By design, this class should intend to handle file opening in `__enter__` or `__init__`, and no where else. Likewise, file closing in a manual `release`, `__del__`, or `__exit__`. All other functionalities acting on the attribute as if it were open, and for safety, checking that it is.